File: biometric/src/main/java/dev/skomlach/biometric/compat/utils/hardware/Android29Hardware.java
package dev.skomlach.biometric.compat.utils.hardware;

import android.annotation.TargetApi;
import android.hardware.biometrics.BiometricManager;
import android.os.Build;

import androidx.annotation.RestrictTo;

import dev.skomlach.biometric.compat.BiometricAuthRequest;
import dev.skomlach.biometric.compat.utils.logging.BiometricLoggerImpl;
import dev.skomlach.common.contextprovider.AndroidContext;
import dev.skomlach.common.misc.Utils;

import static android.hardware.biometrics.BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE;
import static androidx.biometric.BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED;
import static androidx.biometric.BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED;
import static androidx.biometric.BiometricManager.BIOMETRIC_SUCCESS;

@TargetApi(Build.VERSION_CODES.Q)
@RestrictTo(RestrictTo.Scope.LIBRARY)
public class Android29Hardware extends Android28Hardware {

    public Android29Hardware(BiometricAuthRequest authRequest) {
        super(authRequest);
    }

    private int canAuthenticate() {
        int code = BIOMETRIC_ERROR_NO_HARDWARE;
        try {
            BiometricManager biometricManager = AndroidContext.getAppContext().getSystemService(BiometricManager.class);
            if (biometricManager != null) {
                if (Utils.isAtLeastR()) {
                    code = biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_WEAK);
                } else {
                    code = biometricManager.canAuthenticate();
                }
            }
        } catch (Throwable e) {
            BiometricLoggerImpl.e(e);
        }
        BiometricLoggerImpl.e("Android29Hardware.canAuthenticate - " + code);
        return code;
    }

    @Override
    public boolean isAnyHardwareAvailable() {
        int canAuthenticate = canAuthenticate();
        if (canAuthenticate == BIOMETRIC_SUCCESS) {
            return true;
        } else {
            return canAuthenticate != BIOMETRIC_ERROR_NO_HARDWARE && canAuthenticate != BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED;
        }
    }

    @Override
    public boolean isAnyBiometricEnrolled() {
        int canAuthenticate = canAuthenticate();
        if (canAuthenticate == BIOMETRIC_SUCCESS) {
            return true;
        } else {
            return canAuthenticate != BIOMETRIC_ERROR_NONE_ENROLLED && canAuthenticate != BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED;
        }
    }
}

File: biometric/src/main/java/dev/skomlach/biometric/compat/utils/hardware/Android29Hardware.kt
/*
 *  Copyright (c) 2021 Sergey Komlach aka Salat-Cx65; Original project: https://github.com/Salat-Cx65/AdvancedBiometricPromptCompat
 *  All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package dev.skomlach.biometric.compat.utils.hardware

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import androidx.biometric.BiometricManager
import dev.skomlach.biometric.compat.BiometricAuthRequest
import dev.skomlach.biometric.compat.utils.logging.BiometricLoggerImpl.e
import dev.skomlach.common.contextprovider.AndroidContext.appContext
import dev.skomlach.common.misc.Utils.isAtLeastR
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantLock


@TargetApi(Build.VERSION_CODES.Q)

class Android29Hardware(authRequest: BiometricAuthRequest) : Android28Hardware(authRequest) {
    companion object {

        private val lock = ReentrantLock()
        private var cachedCanAuthenticateValue =
            AtomicInteger(BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE)
        private var job: Job? = null
        private var checkStartedTs = 0L


        private fun canAuthenticate(): Int {
            try {
                lock.lock()
                if (job?.isActive == true) {
                    if (System.currentTimeMillis() - checkStartedTs >= TimeUnit.SECONDS.toMillis(30)) {
                        job?.cancel()
                        job = null
                    }
                }
                if (job?.isActive != true) {
                    checkStartedTs = System.currentTimeMillis()
                    job = GlobalScope.launch(Dispatchers.IO) {
                        updateCodeSync()
                    }
                }
                return cachedCanAuthenticateValue.get()
            } finally {
                lock.unlock()
            }
        }

        @SuppressLint("WrongConstant")
        private fun updateCodeSync() {
            var code = BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE
            try {
                var biometricManager: android.hardware.biometrics.BiometricManager? =
                    appContext.getSystemService(
                        android.hardware.biometrics.BiometricManager::class.java
                    )

                if (biometricManager == null) {
                    biometricManager = appContext.getSystemService(
                        Context.BIOMETRIC_SERVICE
                    ) as android.hardware.biometrics.BiometricManager?
                }
                if (biometricManager != null) {
                    code = if (isAtLeastR) {
                        biometricManager.canAuthenticate(android.hardware.biometrics.BiometricManager.Authenticators.BIOMETRIC_WEAK or android.hardware.biometrics.BiometricManager.Authenticators.BIOMETRIC_STRONG)
                    } else {
                        biometricManager.canAuthenticate()
                    }
                }
            } catch (e: Throwable) {
                e(e)
            } finally {
                e("Android29Hardware - canAuthenticate=$code")
                cachedCanAuthenticateValue.set(code)
            }
        }

        init {
            canAuthenticate()
        }
    }

    override val isAnyHardwareAvailable: Boolean
        get() {
            val canAuthenticate = canAuthenticate()
            return if (canAuthenticate == BiometricManager.BIOMETRIC_SUCCESS) {
                true
            } else {
                canAuthenticate != BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE && canAuthenticate != BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED
            }
        }
    override val isAnyBiometricEnrolled: Boolean
        get() {
            val canAuthenticate = canAuthenticate()
            return if (canAuthenticate == BiometricManager.BIOMETRIC_SUCCESS) {
                true
            } else {
                canAuthenticate != BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED && canAuthenticate != BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED
            }
        }
}